test/extensions/filters/http/csrf/csrf_filter_integration_test.cc

```diff
@@ -0,0 +1,200 @@
+#include "test/integration/http_protocol_integration.h"
+
+namespace Envoy {
+namespace {
+const std::string CSRF_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+  shadow_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_FILTER_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_SHADOW_ENABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 0
+      denominator: HUNDRED
+  shadow_enabled:
+    default_value:
+      numerator: 100
+      denominator: HUNDRED
+)EOF";
+
+const std::string CSRF_DISABLED_CONFIG = R"EOF(
+name: envoy.csrf
+config:
+  filter_enabled:
+    default_value:
+      numerator: 0
+      denominator: HUNDRED
+)EOF";
+
+class CsrfFilterIntegrationTest : public HttpProtocolIntegrationTest {
+protected:
+  void testNormalRequest(Http::TestHeaderMapImpl&& request_headers,
+                         const char* expected_response_code) {
+    initialize();
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    auto response = codec_client_->makeRequestWithBody(request_headers, 1024);
+    waitForNextUpstreamRequest();
+    upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, true);
+    response->waitForEndStream();
+
+    ASSERT_TRUE(response->complete());
+    EXPECT_STREQ(expected_response_code, response->headers().Status()->value().c_str());
+  }
+
+  void testInvalidRequest(Http::TestHeaderMapImpl&& request_headers,
+                          const char* expected_response_code) {
+    initialize();
+    codec_client_ = makeHttpConnection(lookupPort("http"));
+    auto response = codec_client_->makeRequestWithBody(request_headers, 1024);
+    response->waitForEndStream();
```

Comment:
I do not see why this is an invalid request. Can you add comments showing how this method is doing what its name indicates? Is it really just that the call-site has the mismatching headers? Maybe you should just factor that out and add them in here?

Reply:
I can modify the method name. It's invalid here in the sense that the CSRF filter is rejecting request since the origins do not match